flink-filesystems/flink-s3-fs-presto/src/main/java/org/apache/flink/fs/s3presto/FlinkS3PrestoFileSystem.java

```diff
@@ -0,0 +1,139 @@
+import java.util.Optional;
+
+/**
+ * {@code FlinkS3PrestoFileSystem} provides custom recursive deletion functionality to work around a
+ * bug in the internally used Presto file system.
+ *
+ * <p>https://github.com/prestodb/presto/issues/17416
+ */
+public class FlinkS3PrestoFileSystem extends FlinkS3FileSystem {
+
+    public FlinkS3PrestoFileSystem(
+            FileSystem hadoopS3FileSystem,
+            String localTmpDirectory,
+            @Nullable String entropyInjectionKey,
+            int entropyLength,
+            @Nullable S3AccessHelper s3UploadHelper,
+            long s3uploadPartSize,
+            int maxConcurrentUploadsPerStream) {
+        super(
+                hadoopS3FileSystem,
+                localTmpDirectory,
+                entropyInjectionKey,
+                entropyLength,
+                s3UploadHelper,
+                s3uploadPartSize,
+                maxConcurrentUploadsPerStream);
+    }
+
+    @Override
+    public boolean delete(Path path, boolean recursive) throws IOException {
+        if (recursive) {
+            deleteRecursively(path);
+        } else {
+            deleteObject(path);
+        }
+
+        return true;
+    }
+
+    private void deleteRecursively(Path path) throws IOException {
+        final FileStatus[] containingFiles =
+                Preconditions.checkNotNull(
+                        listStatus(path),
+                        "Hadoop FileSystem.listStatus should never return null based on its contract.");
+
+        if (containingFiles.length == 0) {
+            deleteObject(path);
+            return;
+        }
+
+        IOException exception = null;
+        for (FileStatus fileStatus : containingFiles) {
+            final Path childPath = fileStatus.getPath();
+
+            try {
+                if (fileStatus.isDir()) {
+                    deleteRecursively(childPath);
+                } else {
+                    deleteObject(childPath);
+                }
+            } catch (IOException e) {
+                exception = ExceptionUtils.firstOrSuppressed(e, exception);
+            }
+        }
+
+        // Presto doesn't hold placeholders for directories itself; therefore, we don't need to
+        // call deleteObject on the directory itself if there were objects with the prefix being
+        // deleted in the initial loop above. This saves us from doing some existence checks on
+        // an not-existing object (i.e. the now empty directory)
+        if (exception != null) {
+            throw exception;
+        }
+    }
+
+    /**
+     * Deletes the object referenced by the passed {@code path}.
+     *
+     * @param path The path referring to the object that shall be deleted.
+     * @throws IOException if an error occurred while deleting the file other than the {@code path}
+     *     referring to a non-empty directory.
+     */
+    private void deleteObject(Path path) throws IOException {
+        if (!exists(path)) {
```

Comment:
I thought we wanted to avoid exist calls? Why not just do the delete call and let the error-handling catch this case.

Reply:
So the reason why we need this method is because `PrestoS3FileSystem#delete` returns `false` both if some error occurred and if the file does not exist?

Reply:
But now you _always_ do the exists check, whereas we could limit this to certain edge-cases.

Reply:
Please document that this method exists because the Presto FS does not allow us to differentiate between deleting a file that doesn't exist (which we'd be fine with) and some random error.

Reply:
You're right. I removed the initial exists call again and added comments